### PR TITLE
Pytest worker shutdown timeout

### LINKED ISSUE #6584: pytest fixture can leave worker running and prevents pytest process to exit
# Checklist

- [x] I have verified that the issue exists against the `master` branch of Celery.
- [ ] This has already been asked to the [discussion group](https://groups.google.com/forum/#!forum/celery-users) first.
- [x] I have read the relevant section in the
  [contribution guide](http://docs.celeryproject.org/en/latest/contributing.html#other-bugs)
  on reporting bugs.
- [x] I have checked the [issues list](https://github.com/celery/celery/issues?q=is%3Aissue+label%3A%22Issue+Type%3A+Bug+Report%22+-label%3A%22Category%3A+Documentation%22)
  for similar or identical bug reports.
- [x] I have checked the [pull requests list](https://github.com/celery/celery/pulls?q=is%3Apr+label%3A%22PR+Type%3A+Bugfix%22+-label%3A%22Category%3A+Documentation%22)
  for existing proposed fixes.
- [x] I have checked the [commit log](https://github.com/celery/celery/commits/master)
  to find out if the bug was already fixed in the master branch.
- [x] I have included all related issues and possible duplicate issues
  in this issue (If there are none, check this box anyway).

## Mandatory Debugging Information

- [ ] I have included the output of ``celery -A proj report`` in the issue.
    (if you are not able to do this, then at least specify the Celery
     version affected).
- [x] I have verified that the issue exists against the `master` branch of Celery.
- [ ] I have included the contents of ``pip freeze`` in the issue.
- [ ] I have included all the versions of all the external dependencies required
  to reproduce this bug.

## Optional Debugging Information

- [ ] I have tried reproducing the issue on more than one Python version
  and/or implementation.
- [ ] I have tried reproducing the issue on more than one message broker and/or
  result backend.
- [ ] I have tried reproducing the issue on more than one version of the message
  broker and/or result backend.
- [ ] I have tried reproducing the issue on more than one operating system.
- [ ] I have tried reproducing the issue on more than one workers pool.
- [ ] I have tried reproducing the issue with autoscaling, retries,
  ETA/Countdown & rate limits disabled.
- [ ] I have tried reproducing the issue after downgrading
  and/or upgrading Celery and its dependencies.

#### Related Issues

- None

#### Possible Duplicates

- None

# Expected Behavior

The `celery_worker` pytest fixture should either wait until all enqueued tasks are processed before stopping the worker or, alternatively, it should raise an exception if the worker fails to terminate within the defined timeout (currently hardcoded to 10s).

# Actual Behavior

If, when the cleanup code of the `celery_worker` pytest fixture is called, the worker is still executing tasks, and those tasks take longer than 10s to terminate, the worker thread is left in running state and the worker will never be stopped, causing the pytest process to never exit.

# Additional information

The code around https://github.com/celery/celery/blob/7288147d65a32b726869ed887d99e4bfd8c070e2/celery/contrib/testing/worker.py#L133 waits for the thread to exit with a timeout of 10 seconds, but the status of the thread once the threading API `.join` method returns is not checked, potentially leaving the thread running.

Because of the thread status not being checked (and the `.join()` method not raising exceptions on timeout), the execution proceeds cleanly, by resetting the `should_terminate` state variable back to `None`, causing the worker thread to stay alive even after it completed all the enqueued tasks.

At a minimum, the fixture should raise an exception if the worker is still alive once the timeout is expired, letting the developer know about the issue and letting him make sure that all tasks have been processed before exiting the test and letting pytest invoke the teardown code. An additional improvement would be to let the timeout be overridable e.g. via another fixture as done for [celery_worker_parameters](https://github.com/celery/celery/blob/7288147d65a32b726869ed887d99e4bfd8c070e2/celery/contrib/pytest.py#L149) or to let the developer indicate that the worker should wait until the current/all pending tasks have been completed before exiting via a more robust method.

### PR DESCRIPTION
**Pull request by GaretJax**
*Note*: Before submitting this pull request, please review our [contributing
guidelines](https://docs.celeryproject.org/en/master/contributing.html).

## Description

Fixes #6584.

This PR is an initial attempt to:
1. Raise an error when a worker thread started by the `celery_worker` fixture does not exit within the predefined timeout
2. Let the user customize the timeout

It is missing tests and there are still a couple TODOs in the code worth checking. I may consider bringing it into a fully mergeable state after an initial round of feedback and depending on time constraints.

### CODE REVIEW

File: celery/contrib/testing/worker.py [hunk 1]
```diff
@@ -121,16 +124,26 @@ def _start_worker_thread(app,
...
+    # TODO: Shall we consider starting the thread with daemon=True
     t = threading.Thread(target=worker.start)
     t.start()
     worker.ensure_started()
     _set_task_join_will_block(False)
 
     yield worker
 
+    # TODO: Shall we check if the worker is still executing tasks, so that
+    # the timeout below is only applied to the shutdown operation?
```
Comment (matusvalo) [inline celery/contrib/testing/worker.py:136]: We don't need to check that. Timeout of `join()` is just upper limit. If worker is not processing tasks, `join()` should return immediately.

File: celery/contrib/testing/worker.py [hunk 2]
```diff
@@ -121,16 +124,26 @@ def _start_worker_thread(app,
         without_gossip=True,
         **kwargs)
 
+    # TODO: Shall we consider starting the thread with daemon=True
```
Comment (matusvalo) [inline celery/contrib/testing/worker.py:127]: +1

**Review by matusvalo — COMMENTED** [review]
Thank you @GaretJax. In general PR looks good. It would be great to finish the PR with setting deamon=True and removing TODO comments.

**Comment by GaretJax** [comment]
Thanks for the review @matusvalo. I just pushed the requested changes.

**Review by auvipy — COMMENTED** [review]
can we test these changes?

**Comment by GaretJax** [comment]
I'm sorry I will not have the bandwidth to add specific tests @auvipy.

**Review by auvipy — APPROVED** [review]

**Comment by auvipy** [comment]
t/integration/test_backend.py::test_AzureBlockBlobBackend::test_crud SKIPPED
  t/integration/test_backend.py::test_AzureBlockBlobBackend::test_get_missing SKIPPED
  t/integration/test_canvas.py::test_link_error::test_link_error_eager PASSED
  t/integration/test_canvas.py::test_link_error::test_link_error PASSED
  t/integration/test_canvas.py::test_link_error::test_link_error_callback_error_callback_retries_eager PASSED
  t/integration/test_canvas.py::test_link_error::test_link_error_callback_retries RERUN
  Error: The operation was canceled.

**Pull request merged**